File: algo/src/main/java/com/hartwig/actin/algo/evaluation/infection/HasSpecificInfection.java
package com.hartwig.actin.algo.evaluation.infection;

import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.algo.datamodel.Evaluation;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.algo.evaluation.EvaluationFactory;
import com.hartwig.actin.algo.evaluation.EvaluationFunction;
import com.hartwig.actin.algo.othercondition.OtherConditionSelector;
import com.hartwig.actin.clinical.datamodel.PriorOtherCondition;
import com.hartwig.actin.doid.DoidModel;

import org.jetbrains.annotations.NotNull;

public class HasSpecificInfection implements EvaluationFunction {

    @NotNull
    private final DoidModel doidModel;
    @NotNull
    private final String doidToFind;

    HasSpecificInfection(@NotNull final DoidModel doidModel, @NotNull final String doidToFind) {
        this.doidModel = doidModel;
        this.doidToFind = doidToFind;
    }

    @NotNull
    @Override
    public Evaluation evaluate(@NotNull PatientRecord record) {
        String doidTerm = doidModel.resolveTermForDoid(doidToFind);

        for (PriorOtherCondition condition : OtherConditionSelector.selectClinicallyRelevant(record.clinical().priorOtherConditions())) {
            for (String doid : condition.doids()) {
                if (doidModel.doidWithParents(doid).contains(doidToFind)) {
                    return EvaluationFactory.unrecoverable()
                            .result(EvaluationResult.PASS)
                            .addPassSpecificMessages("Patient has infection with " + doidTerm)
                            .addPassGeneralMessages("Present " + doidTerm + " infection")
                            .build();
                }
            }
        }

        return EvaluationFactory.unrecoverable()
                .result(EvaluationResult.FAIL)
                .addFailSpecificMessages("Patient has no known infection with " + doidTerm)
                .addFailGeneralMessages("Specific infection(s) not present")
                .build();
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/infection/HasSpecificInfection.kt
package com.hartwig.actin.algo.evaluation.infection

import com.hartwig.actin.algo.evaluation.EvaluationFactory
import com.hartwig.actin.algo.evaluation.EvaluationFunction
import com.hartwig.actin.algo.evaluation.EvaluationLabels
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.algo.Evaluation
import com.hartwig.actin.datamodel.clinical.IcdCode
import com.hartwig.actin.icd.IcdModel

class HasSpecificInfection(
    private val icdModel: IcdModel,
    private val icdCodes: Set<IcdCode>,
    private val term: String,
    private val labels: EvaluationLabels.Infection
) : EvaluationFunction {

    override fun evaluate(record: PatientRecord): Evaluation {
        val matchingComorbidities = icdModel.findInstancesMatchingAnyIcdCode(record.comorbidities, icdCodes)
        val infectionStatus = record.clinicalStatus.infectionStatus
        val hasMatchingInfection = if (infectionStatus?.hasActiveInfection == true) {
            infectionStatus.description?.let { description ->
                description.contains(term) || term.contains(description)
            }
        } else {
            false
        }

        return when {
            matchingComorbidities.fullMatches.isNotEmpty() || hasMatchingInfection == true -> {
                EvaluationFactory.pass(labels.hasSpecificInfectionPass(term.replaceFirstChar(Char::uppercase)))
            }

            hasMatchingInfection == null || matchingComorbidities.mainCodeMatchesWithUnknownExtension.isNotEmpty() -> {
                EvaluationFactory.undetermined(labels.hasSpecificInfectionUndetermined(term))
            }

            else -> EvaluationFactory.fail(labels.hasSpecificInfectionFail(term))
        }
    }
}